portal-bridge/src/types/state/walk_trie_diff.rs

```diff
@@ -0,0 +1,101 @@
+use alloy_primitives::{Bytes, B256};
+use eth_trie::{decode_node, node::Node};
+use hashbrown::HashMap as BrownHashMap;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct AccountProof {
+    pub path: Vec<u8>,
+    pub proof: Vec<Bytes>,
+}
+
+/// This struct takes in a root hash and a hashmap of changed nodes, then you can call an iterator
+/// which will return every proof to gossip
+pub struct WalkTrieDiff {
+    pub changed_nodes: BrownHashMap<Vec<u8>, Vec<u8>>,
+    path_backwards: BrownHashMap<Vec<u8>, (Vec<u8>, u8)>,
+}
+
+impl WalkTrieDiff {
+    pub fn new(root_hash: B256, changed_nodes: BrownHashMap<Vec<u8>, Vec<u8>>) -> Self {
+        let path_backwards = Self::get_paths_backwards(root_hash, &changed_nodes);
+        Self {
+            changed_nodes,
+            path_backwards,
+        }
+    }
+
+    fn get_paths_backwards(
+        root_hash: B256,
+        changed_nodes: &BrownHashMap<Vec<u8>, Vec<u8>>,
+    ) -> BrownHashMap<Vec<u8>, (Vec<u8>, u8)> {
+        let mut path_backwards = BrownHashMap::new();
+        let mut stack = vec![root_hash];
+        while let Some(node_key) = stack.pop() {
+            let encoded_node = changed_nodes
+                .get(&node_key.to_vec())
+                .expect("The stack should only contain nodes that are in the changed nodes");
+
+            let decoded_node = decode_node(&mut encoded_node.as_slice())
+                .expect("Should should only be passing valid encoded nodes");
+
+            match decoded_node {
+                Node::Extension(extension) => {
+                    let extension = extension.read().expect("Reading an extension should work");
+                    if let Node::Hash(hash_node) = &extension.node {
+                        if changed_nodes.contains_key(&hash_node.hash.to_vec()) {
+                            stack.push(hash_node.hash);
+                            path_backwards.insert(
+                                hash_node.hash.to_vec(),
+                                (node_key.to_vec(), extension.prefix.get_data()[0]),
+                            );
+                        }
+                    }
+                }
+                Node::Branch(branch) => {
+                    let branch = branch.read().expect("Reading a branch should work");
+                    for (i, child) in branch.children.iter().enumerate() {
+                        if let Node::Hash(hash_node) = child {
+                            if changed_nodes.contains_key(&hash_node.hash.to_vec()) {
+                                stack.push(hash_node.hash);
+                                path_backwards
+                                    .insert(hash_node.hash.to_vec(), (node_key.to_vec(), i as u8));
+                            }
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        path_backwards
```

Comment:
> I'm not sure if it is possible in the real scenario (definitely it is if there is a bug), but what would happen if there is a node in changed_nodes that is not reachable from provided root?

This wouldn't be possible. changed_nodes includes all the changed trie nodes. So all nodes must be reachable by the root, unless someone did something wrong.

The assumption is we aren't inserting invalid `changed_nodes` into the TrieWalker.

If someone is inserting invalid data, well then of course it would output invalid data as aswell

Garbage in Garbage out